### PR TITLE
External slashing protection

### PR DESCRIPTION
**Pull request by shayzluf**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**
This PR is implementing the proposed external slashing protection 

**Which issues(s) does this PR fix?**

Part of #5806

### CODE REVIEW

File: fuzz/testing/beacon_fuzz_states.go [hunk 1]
```diff
@@ -2,10 +2,11 @@ package testing
 
 import (
 	"fmt"
-	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
-	"github.com/prysmaticlabs/prysm/shared/testutil"
 	"os"
 	"strconv"
+
```
Comment (terencechain) [inline fuzz/testing/beacon_fuzz_states.go:7]:
```suggestion
```

File: nogo_config.json [hunk 2]
```diff
@@ -73,7 +73,7 @@
       "beacon-chain/blockchain/testing/*": "Test-only package",
       "beacon-chain/p2p/sender\\.go": "Libp2p uses time.Now and this file sets a time based deadline in such a way that roughtime cannot be used",
       "beacon-chain/sync/deadlines\\.go": "Libp2p uses time.Now and this file sets a time based deadline in such a way that roughtime cannot be used",
-      "validator/client/grpc_interceptor\\.go": "Uses time.Now() for gRPC duration logging"
```
Comment (terencechain) [inline nogo_config.json:76]: why remove this?

File: shared/logutil/logutil.go [hunk 3]
```diff
@@ -23,3 +27,24 @@ func ConfigurePersistentLogging(logFileName string) error {
 	logrus.Info("File logging initialized")
 	return nil
 }
+
+// LogDebugRequestInfoUnaryInterceptor this method logs the gRPC backend as well as request duration when the log level is set to debug
+// or higher.
+func LogDebugRequestInfoUnaryInterceptor(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
```
Comment (terencechain) [inline shared/logutil/logutil.go:33]: What is this?

File: nogo_config.json (hunk 2 above)
Comment (shayzluf) [reply nogo_config.json:76]: Moved it t ol shared

File: shared/logutil/logutil.go (hunk 3 above)
Comment (shayzluf) [reply shared/logutil/logutil.go:33]: Grpc logging

File: fuzz/testing/beacon_fuzz_states.go (hunk 1 above)
Comment (prestonvanloon) [reply fuzz/testing/beacon_fuzz_states.go:7]: Why? This new line should be here

File: slasher/detection/listeners.go [hunk 4]
```diff
@@ -9,9 +9,9 @@ package detection
 import (
 	"context"
 
-	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/shared/blockutil"
+
```
Comment (prestonvanloon) [inline slasher/detection/listeners.go:13]:
```suggestion
```

File: validator/client/service.go [hunk 5]
```diff
@@ -6,6 +6,10 @@ import (
 	"context"
 	"strings"
 
+	slashingprotection "github.com/prysmaticlabs/prysm/validator/slashing-protection"
+
+	"github.com/prysmaticlabs/prysm/shared/logutil"
+
```
Comment (prestonvanloon) [inline validator/client/service.go:12]: Please clean up imports

File: validator/flags/flags.go [hunk 6]
```diff
@@ -23,6 +23,17 @@ var (
 		Name:  "tls-cert",
 		Usage: "Certificate for secure gRPC. Pass this and the tls-key flag in order to use gRPC securely.",
 	}
+	// SlasherRPCProviderFlag defines a slasher node RPC endpoint.
+	SlasherRPCProviderFlag = &cli.StringFlag{
+		Name:  "slasher-rpc-provider",
+		Usage: "Slasher node RPC provider endpoint",
+		Value: "localhost:4000",
```
Comment (prestonvanloon) [inline validator/flags/flags.go:30]: The default is the same address as the beacon rpc?

Comment (shayzluf) [reply validator/flags/flags.go:30]: Missed it, thanks

File: shared/logutil/logutil.go (hunk 3 above)
Comment (rauljordan) [reply shared/logutil/logutil.go:33]: Could we add a bit more descriptive comments for this function or make the name simpler?

Comment (shayzluf) [reply shared/logutil/logutil.go:33]: changed the method name

File: validator/client/service.go (hunk 5 above)
Comment (shayzluf) [reply validator/client/service.go:12]: thanks

File: validator/flags/flags.go (hunk 6 above)
Comment (shayzluf) [reply validator/flags/flags.go:30]: changed to 4001

File: validator/client/validator_propose.go [hunk 7]
```diff
@@ -42,6 +43,16 @@ var (
 			"pubkey",
 		},
 	)
+	validatorProposeFailVecSlasher = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "validator",
+			Name:      "failed_proposals_external_slasher",
```
Comment (terencechain) [inline validator/client/validator_propose.go:49]: please follow prometheus naming standard:
https://prometheus.io/docs/practices/naming/

File: validator/client/validator_attest.go [hunk 8]
```diff
@@ -136,7 +146,23 @@ func (v *validator) SubmitAttestation(ctx context.Context, slot uint64, pubKey [
 		AggregationBits: aggregationBitfield,
 		Signature:       sig,
 	}
-
+	if v.protector != nil {
```
Comment (terencechain) [inline validator/client/validator_attest.go:149]: I'd add the `SlasherProtection` check here for readability

**Comment by terencechain** [comment]
> @terencechain I think the scope of the PR is fine, it mostly has logic self-contained towards adding a slasher client to the validator. Just had a small comment

This does violate the one PR, one single responsibility type of principle because it has implemented both server and client responsibilities and connect them for end to end usage.
1.) Implemented server side slasher protection interface
2.) Implemented validator grpc client
3.) Implemented grpc client to slasher protection and gated via a feature flag + metrics.. etc

We can always make an exception out of this but it moves us to a gray area and raises questions such what is good practice and when a scope is too big

I'll let this marinate for a few hours and if no one has as a strong opinion, I'll dismiss my request

File: validator/client/validator_propose.go (hunk 7 above)
Comment (nisdas) [reply validator/client/validator_propose.go:49]: +1

**Comment by shayzluf** [comment]
@terencechain sorry it became so big i had #5807 and built this PR off it. i didn't mention it enough in the PR and got most of the reviews on this one. decided to close #5807 and just apply all the feedbacks here

File: validator/client/validator_attest.go (hunk 8 above)
Comment (shayzluf) [reply validator/client/validator_attest.go:149]: just thought that this protector can be implemented differently then by `SlasherProtection` wdyt?

File: validator/slashing-protection/slasher.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
+package slashingprotection
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply validator/slashing-protection/slasher.go:1]: thanks renamed

**Comment by terencechain** [comment]
> @terencechain sorry it became so big i had #5807 and built this PR off it. i didn't mention it enough in the PR and got most of the reviews on this one. decided to close #5807 and just apply all the feedbacks here

Not a problem, the changes on the server was rather minimal. Thanks for the  great work as always. Happy to click approve on this soon!

File: validator/client/validator_propose.go (hunk 7 above)
Comment (shayzluf) [reply validator/client/validator_propose.go]: changed to 
```prometheus.CounterOpts{
			Name: "validator_proposals_rejected_total",
			Help: "Count the block proposals rejected by protector interface.",
		},
```
```prometheus.CounterOpts{
			Name: "validator_attestations_rejected_total",
			Help: "Count the attestations rejected by protector interface.",
		},
```
is this ok

File: slasher/rpc/server.go [hunk 10]
```diff
@@ -86,16 +86,15 @@ func (ss *Server) IsSlashableAttestation(ctx context.Context, req *ethpb.Indexed
...
-	if err := ss.slasherDB.SaveIndexedAttestation(ctx, req); err != nil {
-		log.WithError(err).Error("Could not save indexed attestation")
-		return nil, status.Errorf(codes.Internal, "Could not save indexed attestation: %v: %v", req, err)
-	}
 	slashings, err := ss.detector.DetectAttesterSlashings(ctx, req)
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "Could not detect attester slashings for attestation: %v: %v", req, err)
 	}
 	if len(slashings) < 1 {
+		if err := ss.slasherDB.SaveIndexedAttestation(ctx, req); err != nil {
```
Comment (nisdas) [inline slasher/rpc/server.go:94]: why put it here, shouldn't all attestations be saved

File: validator/client/validator_attest.go (hunk 8 above)
Comment (nisdas) [reply validator/client/validator_attest.go:149]: its probably better to emphasize that we need the flag that we need this to work. On an initial look it wouldn't be clear

**Pull request merged**